wolfssl/src/ssl.rs

```diff
@@ -1176,8 +1168,9 @@ impl<IOCB: IOCallbacks> Session<IOCB> {
         debug_assert!(!secret.is_null());
         debug_assert!(!ctx.is_null());
 
-        // let keylog = Box::from_raw(ctx as *mut Arc<dyn Tls13SecretCallbacks>);
-        let secret_cb = &*(ctx as *mut Tls13SecretCallbacksArg);
+        // SAFETY: We have only one strong reference in `self.secret_cb`
+        // Leak the Box pointer, so that it will not be freed while this function return
+        let secret_cb = Box::leak(Box::from_raw(ctx as *mut Tls13SecretCallbacksArg));
```

Comment:
Can you expand (here or perhaps in the commit message) on what the issue was with the previous code using `Arc::as_ptr` etc? I expected it would work and `secret_cb` here would have type `&Arc<Tls13SecretCallbacksArg>` which I thought would be fine.

Since this is obviously pretty subtle maybe `wolfssl/tests/ping_pong.rs` could register the cb when running in debug mode and `eprintln` it?

Reply:
I do not know for sure.
This is my theory. I have tried some examples but could not see it 100%

But based on my reading, `Arc::as_ptr()` will return raw_pointer of data and in this case  its a `<dyn Callback>`. And it is a fat pointer, since its a trait object.

```rust
let secret_cb = Arc::as_ptr(&secret_cb) as *mut c_void;
```

And when we reconstruct it using :
```rust
let secret_cb =  &*(ctx as *mut Tls13SecretCallbacksArg);
```

We actually treat the pointer as the pointer to the Arc and not the actual data (trait object). So when it tries to dereference, and application crashes.

I have also tried `let secret_cb =  Arc::from_raw(ctx as *const dyn Tls13SecretCallbacks);`. But compiler does not like the type cast of `* mut c_void` to `*const dyn Tls13SecretCallbacks`.

So I come up with putting the Arc inside the box and send the box pointer as context pointer. It avoids the above fat pointer issue. While reconstruction, I just recreated a Box from context pointer and leak it at same time to avoid memory free at the end of the call.

Reply:
I'll have a play with this, leaking things should not be needed.

Reply:
I figured it out and pushed it as a commit.